### PR TITLE
Introduce CIS2

### LINKED ISSUE #23: Introduce CIS2 token standard
The new features in smart contract V1 enables some simplification of the token standard CIS1.

In smart contract V1 functions now have output, which deprecates the callback pattern of the query functions in CIS1.

### PR DESCRIPTION
**Pull request by limemloh**
## Purpose

Introducing a new token standard CIS2 which takes advantage of new functionality introduced in smart contract v1.
Closes #23 

## Changes

- Introduce CIS2 token standard
- Mark CIS1 as deprecated by CIS2

### CODE REVIEW

**Comment by limemloh** [comment]
Note that I have added a section with the changes from CIS1 at the bottom.

File: source/CIS/cis-2.rst [hunk 1]
```diff
@@ -0,0 +1,790 @@
...
+Token standards such as ERC721 and ERC1155 both use a 256-bit unsigned integer (32 bytes) for the token ID, to support using something like a SHA256 hash for the token ID.
+But in the case where the token ID have no significance other than a simple identifier, smaller sized token IDs can reduce energy costs.
+This is why we chose to let the first byte indicate the size of the token ID, meaning a token ID can vary between 1 byte and 256 bytes, resulting in more than 10^614 possible token IDs.
+
+Only batched transfers
+----------------------
+
+The specification only has a ``transfer`` smart contract function which takes list of transfer and no function for a single transfer.
+This will result in lower energy cost compared to multiple contract calls and only introduces a small overhead for single transfers.
+The reason for not also including a single transfer function is to have smaller smart contract modules, which in turn leads to saving cost on every function call.
```
Comment (abizjak) [inline source/CIS/cis-2.rst:702]: Can you also add that there is "only transferFrom" (which is the transfer) since that was apparently not clear.

File: source/CIS/cis-2.rst [hunk 2]
```diff
@@ -0,0 +1,790 @@
...
+
+  The smart contract receive hook function could be called by any smart contract and it is up to the integrating contract whether to trust the token contract.
+
+Receive hook function callback argument
+---------------------------------------
+
+The name of the receive hook function called on a smart contract receiving tokens is supplied as part of the parameter.
+This allows for a smart contract integrating with a token smart contract to have multiple hooks and leave it to the caller to know which hook they want to trigger.
+
+Another technical reason is that the name of the smart contract is part of the smart contract receive function name, which means the specification would include a requirement of the smart contract name for others to integrate reliably.
```
Comment (abizjak) [inline source/CIS/cis-2.rst:750]: As pointed out above, this rationale no longer applies. So this paragraph needs to be removed.

File: source/CIS/cis-2.rst [hunk 3]
```diff
@@ -0,0 +1,790 @@
...
+
+.. _CIS-2-event-tokenMetadata:
+
+``TokenMetadataEvent``
+^^^^^^^^^^^^^^^^^^^^^^
+
+The event to log when setting the metadata url for a token type.
+
+It consists of a token ID and a URL (:rfc:`3986`) for the location of the metadata for this token type with an optional SHA256 checksum of the content.
+Logging the ``TokenMetadataEvent`` event again with the same token ID, is used to update the metadata location and only the most recently logged token metadata event for certain token id should be used to get the token metadata.
```
Comment (Bargsteen) [inline source/CIS/cis-2.rst:475]:
```suggestion
Logging the ``TokenMetadataEvent`` event again with the same token ID, is used to update the metadata location and only the most recently logged token metadata event for a certain token id should be used to get the token metadata.
```

File: source/CIS/cis-2.rst [hunk 4]
```diff
@@ -0,0 +1,790 @@
...
+
+The ``TokenMetadataEvent`` event is serialized as: first a byte with the value of 251, followed by the token ID :ref:`CIS-2-TokenID` (``id``) and then a :ref:`CIS-2-MetadataUrl` (``metadata``)::
```
Comment (Bargsteen) [inline source/CIS/cis-2.rst:477]:
```suggestion
The ``TokenMetadataEvent`` event is serialized as: first a byte with the value of 251, followed by the token ID :ref:`CIS-2-TokenID` (``id``), and then a :ref:`CIS-2-MetadataUrl` (``metadata``)::
```

File: source/CIS/cis-2.rst [hunk 5]
```diff
@@ -0,0 +1,790 @@
...
+
+  TokenMetadataEvent ::= (251: Byte) (id: TokenID) (metadata: MetadataUrl)
+
+.. note::
+
+  Be aware of the limit on the number of logs per smart contract function call, which currently is 64, and also the byte size limit on each logged event, which currently is 512 bytes.
```
Comment (Bargsteen) [inline source/CIS/cis-2.rst:483]: Log limit

File: source/CIS/cis-2.rst [hunk 6]
```diff
@@ -0,0 +1,790 @@
...
+
+Rejecting using an error code from the table above MUST only occur in a situation as described in the corresponding error description.
+
+The smart contract implementing this specification MAY introduce custom error codes other than the ones specified in the table above.
+
+
+Token metadata JSON
+-------------------
+
+The token metadata is stored off chain and MUST be a JSON (:rfc:`8259`) file.
```
Comment (Bargsteen) [inline source/CIS/cis-2.rst:519]:
```suggestion
The token metadata is stored off-chain and MUST be a JSON (:rfc:`8259`) file.
```

File: source/CIS/cis-2.rst [hunk 7]
```diff
@@ -0,0 +1,790 @@
...
+    - A URL to the token asset.
+  * - ``assets`` (optional)
+    - JSON array of Token metadata JSON objects
+    - Collection of assets.
+  * - ``attributes`` (optional)
+    - JSON array of Attribute JSON objects
+    - Assign a number of attributes to the token type.
+      Attributes can be used to include extra information about the token type.
+  * - ``localization`` (optional)
+    - JSON object with locales as field names (:rfc:`5646`) and field values are URL JSON object to JSON files.
```
Comment (Bargsteen) [inline source/CIS/cis-2.rst:559]: Perhaps? If not, then rephrase, as it is not exactly clear.
```suggestion
    - JSON object with locales as field names (:rfc:`5646`) and field values are URL JSON objects linking to JSON files.
```

File: source/CIS/cis-2.rst [hunk 8]
```diff
@@ -0,0 +1,790 @@
...
+    - URL's to JSON files with localized token metadata.
```
Comment (Bargsteen) [inline source/CIS/cis-2.rst:560]:
```suggestion
    - URLs to JSON files with localized token metadata.
```

File: source/CIS/cis-2.rst [hunk 9]
```diff
@@ -0,0 +1,790 @@
...
+
+Optionally a SHA256 hash of the JSON file can be logged with the TokenMetadata event for checking integrity.
+Since the metadata json file could contain URLs, a SHA256 hash can optionally be associated with the URL.
```
Comment (Bargsteen) [inline source/CIS/cis-2.rst:563]:
```suggestion
Since the metadata JSON file could contain URLs, a SHA256 hash can optionally be associated with the URL.
```

File: source/CIS/cis-2.rst [hunk 10]
```diff
@@ -0,0 +1,790 @@
...
+    }
+  }
+
+The danish localization JSON file could be:
+
+.. code-block:: json
+
+  {
+    "name": "Bibi - Ryan katten",
+    "description": "Ryan katte er ensomme væsner, som rejser rundt i galaxen søgende efter deres forfædre og sande fortid"
```
Comment (Bargsteen) [inline source/CIS/cis-2.rst:672]: ?
```suggestion
    "description": "Ryan katte er ensomme væsner, som rejser rundt i galaxen søgende efter deres forfædre og sande arv"
```

File: source/CIS/cis-2.rst [hunk 11]
```diff
@@ -0,0 +1,790 @@
...
+- Each logged event is limited to 0.5 KiB.
+- The number of logged events is limited to 64 per contract function invocation.
+- The total size of the smart contract module is limited to 512 KiB.
+
+Decisions and rationale
+=======================
+
+In this section we point out some of the differences from other popular token standards found on other blockchains, and provide reasons for deviating from them in CIS2.
+
+Token ID bytes instead an integer
```
Comment (Bargsteen) [inline source/CIS/cis-2.rst:690]:
```suggestion
Token ID bytes instead of integers
```

File: source/CIS/cis-2.rst [hunk 12]
```diff
@@ -0,0 +1,790 @@
...
+---------------------------------
+
+Token standards such as ERC721 and ERC1155 both use a 256-bit unsigned integer (32 bytes) for the token ID, to support using something like a SHA256 hash for the token ID.
+But in the case where the token ID have no significance other than a simple identifier, smaller sized token IDs can reduce energy costs.
+This is why we chose to let the first byte indicate the size of the token ID, meaning a token ID can vary between 1 byte and 256 bytes, resulting in more than 10^614 possible token IDs.
```
Comment (Bargsteen) [inline source/CIS/cis-2.rst:695]:
```suggestion
This is why we chose to let the first byte indicate the size of the token ID, meaning a token ID can vary between 1 byte and 256 bytes. The latter allows more than 10^614 possible token IDs.
```

Or perhaps:
```suggestion
This is why we chose to let the first byte indicate the size of the token ID, meaning a token ID can vary between 1 byte and 256 bytes, allowing up to more than 10^614 possible token IDs.
```

File: source/CIS/cis-2.rst [hunk 13]
```diff
@@ -0,0 +1,790 @@
...
+
+The specification only has a ``transfer`` smart contract function which takes list of transfer and no function for a single transfer.
+This will result in lower energy cost compared to multiple contract calls and only introduces a small overhead for single transfers.
+The reason for not also including a single transfer function is to have smaller smart contract modules, which in turn leads to saving cost on every function call.
+
+No explicit authorization
+-------------------------
+
+The specification does not mandate any authorization scheme and one might expect a requirement for the owner and operators being authorized to transfer tokens.
+This is very much intentional and the reasoning for this is to keep the specification focused on the interface for transferring token ownership with as few restrictions as possible.
```
Comment (Bargsteen) [inline source/CIS/cis-2.rst:708]:
```suggestion
This is intentional and the reason for this is to keep the specification focused on the interface for transferring token ownership with as few restrictions as possible.
```

File: source/CIS/cis-2.rst [hunk 14]
```diff
@@ -0,0 +1,790 @@
...
+
+Having a requirement that only owners and operators can transfer would prevent introducing any other authorization scheme on top of this specification.
+
+Adding a requirement for owners and operators being authorized to transfer tokens would prevent introducing custom contract logic rejecting transfers, such as limiting the daily transfers, temporary token lockups or non-transferrable tokens.
+
+Instead, this specification includes a requirement to ensure transfers by operators are executed as if they are sent by the owner, meaning whenever a token owner is authorized, so is an operator of the owner.
+
+Most contracts implementing this specification should probably add some authorization and not have anyone being able to transfer any token, but this is not in the scope of this standard.
```
Comment (Bargsteen) [inline source/CIS/cis-2.rst:716]: `not have anyone being able to transfer any token` is not so clear. Could you rephrase it? 😊

File: source/CIS/cis-2.rst [hunk 15]
```diff
@@ -0,0 +1,790 @@
...
+---------------------
+
+The specification requires a token receive hook to be called on a smart contract receiving tokens, this will in some cases prevent mistakes such as sending tokens to smart contracts which do not define behavior for receiving tokens.
+These token could then be lost forever.
+
+The reason for this not being optional is to allow other smart contracts which integrate with a token smart contract to rely on this for functionality.
+
+.. warning::
+
+  The smart contract receive hook function could be called by any smart contract and it is up to the integrating contract whether to trust the token contract.
```
Comment (Bargsteen) [inline source/CIS/cis-2.rst:742]: It could be any caller, not just a token contract.
```suggestion
  The smart contract receive hook function could be called by any smart contract and it is up to the integrating contract whether to trust the caller.
```

File: source/CIS/cis-2.rst [hunk 16]
```diff
@@ -0,0 +1,790 @@
...
+
+The name of the receive hook function called on a smart contract receiving tokens is supplied as part of the parameter.
+This allows for a smart contract integrating with a token smart contract to have multiple hooks and leave it to the caller to know which hook they want to trigger.
+
+Another technical reason is that the name of the smart contract is part of the smart contract receive function name, which means the specification would include a requirement of the smart contract name for others to integrate reliably.
+
+No sender hook function
+-----------------------
+
+The FA2 token standard found on Tezos, allows for a hook function to be called on a smart contract sending tokens, such that the contract could reject the transfer on some criteria.
```
Comment (Bargsteen) [inline source/CIS/cis-2.rst:755]:
```suggestion
The FA2 token standard found on Tezos allows for a hook function to be called on a smart contract sending tokens, such that the contract can reject the transfer on some criteria.
```

File: source/CIS/cis-2.rst [hunk 17]
```diff
@@ -0,0 +1,790 @@
...
+This seems to only make sense, if some operator is transferring tokens from a contract, in which case the sender smart contract might as well contain the logic to transfer the tokens and trigger this directly.
```
Comment (Bargsteen) [inline source/CIS/cis-2.rst:756]:
```suggestion
This seems to only make sense if some operator is transferring tokens from a contract, in which case the sender smart contract might as well contain the logic to transfer the tokens and trigger this directly.
```

File: source/CIS/cis-2.rst [hunk 18]
```diff
@@ -0,0 +1,790 @@
...
+Adding SHA256 checksum for token metadata event
+-----------------------------------------------
+
+A token can optionally include a SHA256 checksum when logging the token metadata event, this is to ensure the integrity of the token metadata.
+This checksum can be updated by logging a new event.
+
+Differences from CIS1
+---------------------
+
+Only the query functions :ref:`CIS-2-functions-balanceOf`, :ref:`CIS-2-functions-operatorOf` and :ref:`CIS-2-functions-tokenMetadata` are different from CIS1.
```
Comment (Bargsteen) [inline source/CIS/cis-2.rst:779]:
```suggestion
Only the query functions :ref:`CIS-2-functions-balanceOf`, :ref:`CIS-2-functions-operatorOf`, and :ref:`CIS-2-functions-tokenMetadata` differ from CIS1.
```

File: source/CIS/cis-2.rst [hunk 19]
```diff
@@ -0,0 +1,790 @@
...
+The query functions in CIS1 use a callback pattern to output the result of a query, but starting from Concordium smart contract v1; a smart contract receive function can output bytes back to the function caller.
```
Comment (Bargsteen) [inline source/CIS/cis-2.rst:780]:
```suggestion
The query functions in CIS1 use a callback pattern to output the result of a query, but starting from Concordium smart contract v1 a smart contract receive function can output bytes back to the invoker.
```

File: source/CIS/cis-2.rst [hunk 20]
```diff
@@ -0,0 +1,790 @@
...
+CIS2 uses this output instead of a callback pattern to return the query result.
+Using output instead of callbacks require less energy and will reduce contract code needed for querying.
```
Comment (Bargsteen) [inline source/CIS/cis-2.rst:782]:
```suggestion
Using output instead of callbacks requires less energy and will reduce the contract code needed for querying.
```

File: source/CIS/cis-2.rst [hunk 21]
```diff
@@ -0,0 +1,790 @@
...
+
+In CIS1 the callback result includes the corresponding query to ease the use with callback pattern, which is not needed for the output result in CIS2 query functions.
```
Comment (Bargsteen) [inline source/CIS/cis-2.rst:784]:
```suggestion
In CIS1 the callback result includes the corresponding query to ease the use of the callback pattern. The query information is not needed in the output of CIS2 query functions.
```

File: source/CIS/cis-2.rst [hunk 22]
```diff
@@ -0,0 +1,790 @@
...
+Instead the result are required to be the same length and order as the queries.
```
Comment (Bargsteen) [inline source/CIS/cis-2.rst:785]:
```suggestion
Instead, the results are required to be the same length and order as the queries.
```

File: source/CIS/cis-2.rst [hunk 23]
```diff
@@ -0,0 +1,790 @@
...
+
+In CIS2 smart contract functions are not required to fail with a specific error code as in CIS1. This is to allow receive functions to fail early for reason specific to the implementation such as authorization or serialization.
+
+Prior to smart contract version 1 invoking another smart contract required to know the contract name as well as the contract address and endpoint.
```
Comment (Bargsteen) [inline source/CIS/cis-2.rst:789]:
```suggestion
Prior to smart contract version 1, invoking another smart contract required knowing the contract name as well as the contract address and endpoint.
```

**Review by Bargsteen — CHANGES_REQUESTED** [review]
Nice work. A bit of grammar and leftovers from CIS-1 😊

File: source/CIS/cis-1.rst [hunk 24]
```diff
@@ -9,6 +11,12 @@ CIS-1: Concordium Token Standard
      - Sep 22, 2021
    * - Final
      - Nov 1, 2021
+   * - Supported Smart Contract versions
+     - 0+
```
[the remark this answers is not in the document]
Comment (limemloh) [reply source/CIS/cis-1.rst:15]: I have added a short sentence and specified the protocol version as well

File: source/CIS/cis-2.rst [hunk 25]
```diff
@@ -0,0 +1,790 @@
...
+  }
+
+Smart contract limitations
+==========================
+
+A number of limitations are important to be aware of:
+
+- Smart contract function parameters are limited to 1 KiB.
+- Each logged event is limited to 0.5 KiB.
+- The number of logged events is limited to 64 per contract function invocation.
```
[the remark this answers is not in the document]
Comment (limemloh) [reply source/CIS/cis-2.rst:682]: That is also how I would interprete it.

per contract _function_ invocation.

But it might be better to use _invoke calls_ since we have moved from message passing.
What do you think?

File: source/CIS/cis-2.rst (hunk 1 above)
Comment (limemloh) [reply source/CIS/cis-2.rst:702]: Sure thing.
FYI transfer in CIS2 is even more general than `transferFrom` and `batchTransferFrom` in ERC721 and ERC1155, because these only have a single sender and receiver, and only allow you to transfer a list of tokens and amounts.

File: source/CIS/cis-2.rst [hunk 26]
```diff
@@ -0,0 +1,790 @@
...
+
+Additional bytes to include in a transfer, which can be used to add additional parameters for the transfer function call.
+
+It is serialized as: the first 2 bytes encode the length (``n``) of the data, followed by this many bytes for the data (``data``)::
+
+  AdditionalData ::= (n: Byte²) (data: Byteⁿ)
+
+.. note::
+
+  This type is passed in a parameter for smart contract function calls.
```
[the remark this answers is not in the document]
Comment (limemloh) [reply source/CIS/cis-2.rst]: The type is passed as part of a parameter and not as "whole" parameter, which is why the use of _in_.
I have tried to change the phrasing to emphasize this now.

File: source/CIS/cis-2.rst [hunk 27]
```diff
@@ -0,0 +1,790 @@
...
+- The contract function MUST NOT add or remove any operator for any address.
+- The contract function MUST reject if any of the queries fail:
+
+  - A query MUST fail if the token ID is unknown.
+
+Logged events
+-------------
+
+The idea of the logged events for this specification is for off-chain applications to be able to track balances and operators without knowledge of the contract-specific implementation details.
+For this reason it is important to log events in any functionality of the token contract which modifies balances or operators.
```
[the remark this answers is not in the document]
Comment (limemloh) [reply source/CIS/cis-2.rst:411]: I would find this a bit unclear, since the sentence have both a log event and a contract event.
But I do see the problem with the current phrasing.

[the remark this answers is not in the document]
Comment (limemloh) [reply source/CIS/cis-2.rst:411]: What about:
```
For this reason, it is important for the token contract to log the appropiate event anytime it modifies balances or operators.
```

File: source/CIS/cis-2.rst [hunk 28]
```diff
@@ -776,15 +793,18 @@ This checksum can be updated by logging a new event.
 Differences from CIS1
 ---------------------
 
-Only the query functions :ref:`CIS-2-functions-balanceOf`, :ref:`CIS-2-functions-operatorOf` and :ref:`CIS-2-functions-tokenMetadata` are different from CIS1.
-The query functions in CIS1 use a callback pattern to output the result of a query, but starting from Concordium smart contract v1; a smart contract receive function can output bytes back to the function caller.
+The query functions :ref:`CIS-2-functions-balanceOf`, :ref:`CIS-2-functions-operatorOf`, and :ref:`CIS-2-functions-tokenMetadata` differ from CIS1.
+The query functions in CIS1 use a callback pattern to output the result of a query, but starting from Concordium smart contract v1; a smart contract receive function can output bytes back to the invoker.
```
[the remark this answers is not in the document]
Comment (limemloh) [reply source/CIS/cis-2.rst:797]:
```suggestion
The query functions in CIS1 use a callback pattern to output the result of a query. However, starting from Concordium smart contract v1, a smart contract receive function can output bytes back to the invoker.
```

File: source/CIS/cis-2.rst [hunk 29]
```diff
@@ -669,51 +674,65 @@ The danish localization JSON file could be:
...
+
 Smart contract limitations
 ==========================
 
 A number of limitations are important to be aware of:
 
 - Smart contract function parameters are limited to 1 KiB.
 - Each logged event is limited to 0.5 KiB.
-- The number of logged events is limited to 64 per contract function invocation.
+- The number of logged events is limited to 64 in a contract function invocation. Invoking another contract function will allow another 64 events to be logged.
```
[the remark this answers is not in the document]
Comment (limemloh) [reply source/CIS/cis-2.rst:689]: I agree it is not clear...
I have tried to rephrase it

File: source/CIS/cis-2.rst [hunk 30]
```diff
@@ -0,0 +1,811 @@
...
+
+  Token IDs can be as small as a single byte (by setting the first byte to the value 0) or as big as 256 bytes leaving more than 10^614 possible token IDs.
+  The token ID could be an encoding of a small text string or some checksum hash, but to save energy it is still recommended to use small token IDs if possible.
+
+.. _CIS-2-TokenAmount:
+
+``TokenAmount``
+^^^^^^^^^^^^^^^
+
+An amount of a token type is an unsigned integer up to 2^256.
```
Comment (abizjak) [inline source/CIS/cis-2.rst:65]:
```suggestion
An amount of a token type is an unsigned integer up to (but not including) 2^256.
```
It's technically up to `2^256 - 1`

File: source/CIS/cis-2.rst [hunk 31]
```diff
@@ -0,0 +1,811 @@
...
+- The contract function MUST NOT add or remove any operator for any address.
+- The contract function MUST reject if any of the queries fail:
+
+  - A query MUST fail if the token ID is unknown.
+
+Logged events
+-------------
+
+The idea of the logged events for this specification is for off-chain applications to be able to track balances and operators without knowledge of the contract-specific implementation details.
+For this reason, it is important for the token contract to log the appropriate event, anytime modifications of balances or operators are made.
```
Comment (abizjak) [inline source/CIS/cis-2.rst:413]:
```suggestion
For this reason, it is important for the token contract to log the appropriate event, any time modifications of balances or operators are made.
```

File: source/CIS/cis-2.rst [hunk 32]
```diff
@@ -0,0 +1,811 @@
...
+
+Smart contract limitations
+==========================
+
+A number of limitations are important to be aware of:
+
+- Smart contract function parameters are limited to 1 KiB.
+- Each logged event is limited to 0.5 KiB.
+- The number of logged events is limited to 64 in a contract function invocation.
+  However, invoking another contract function during a contract call, resets the count and allows another 64 events to be logged.
```
Comment (abizjak) [inline source/CIS/cis-2.rst:690]: This is a strange phrasing. 
```suggestion
- The number of logged events is limited to 64 without invoking another contract.
  However invoking another contract function during a contract call resets the count and allows another 64 events to be logged.
```

File: source/CIS/cis-2.rst [hunk 33]
```diff
@@ -0,0 +1,811 @@
...
+-----------------------------------------------
+
+A token can optionally include a SHA256 checksum when logging the token metadata event, this is to ensure the integrity of the token metadata.
+This checksum can be updated by logging a new event.
+
+Differences from CIS1
+---------------------
+
+The query functions :ref:`CIS-2-functions-balanceOf`, :ref:`CIS-2-functions-operatorOf`, and :ref:`CIS-2-functions-tokenMetadata` differ from CIS1.
+The query functions in CIS1 use a callback pattern to output the result of a query. However, starting from Concordium smart contract v1, a smart contract receive function can output bytes back to the invoker.
```
Comment (abizjak) [inline source/CIS/cis-2.rst:798]:
```suggestion
The query functions in CIS1 use a callback pattern to output the result of a query. However, starting from Concordium smart contract version 1, a smart contract receive function can return values back to the invoker.
```

**Review by abizjak — APPROVED** [review]

File: source/CIS/cis-2.rst (hunk 30 above)
Comment (limemloh) [reply source/CIS/cis-2.rst:65]:
```suggestion
An amount of a token type is an unsigned integer up to 2^256 - 1.
```

File: source/CIS/cis-2.rst (hunk 29 above)
[the remark this answers is not in the document]
Comment (Bargsteen) [reply source/CIS/cis-2.rst]: I think this is OK now.

File: source/CIS/cis-2.rst [hunk 34]
```diff
@@ -0,0 +1,811 @@
...
+
+  The specification does not prevent adding more fine-grained authorization, such as token-level operators.
+
+Receive hook function
+---------------------
+
+The specification requires a token receive hook to be invoked on a smart contract receiving tokens, this will in some cases prevent mistakes such as sending tokens to smart contracts which do not define behavior for receiving tokens.
+These token could then be lost forever.
+
+The reason for this not being optional is to allow other smart contracts which integrate with a token smart contract to rely on this for functionality.
```
Comment (Bargsteen) [inline source/CIS/cis-2.rst:758]:
```suggestion
The reason for this not being optional is to allow other smart contracts, which integrate with a token smart contract, to rely on this for functionality.
```

File: source/CIS/cis-2.rst [hunk 35]
```diff
@@ -0,0 +1,811 @@
...
+
+.. warning::
+
+  The smart contract receive hook function can be called by any smart contract and it is up to the integrating contract whether to trust the token contract.
```
Comment (Bargsteen) [inline source/CIS/cis-2.rst:762]:
```suggestion
  The smart contract receive hook function can be called by any account or smart contract. It is up to the integrating contract whether it should trust the caller or not.
```

**Review by Bargsteen — APPROVED** [review]
Just found two minor issues. Other than that, it looks good 👍
